Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp

```diff
@@ -0,0 +1,205 @@
+  /// @brief Detector configuration struct : contains the configuration for each detector
+  ///
+  /// The configuration can be saved in a json file and loaded from there.
+  ///
+  struct DetectorConfig {
+    int hitsScoreWeight = 0;
+    int holesScoreWeight = 0;
+    int outliersScoreWeight = 0;
+    int otherScoreWeight = 0;
+
+    std::size_t minHits = 0;
+    std::size_t maxHits = 0;
+    std::size_t maxHoles = 0;
+    std::size_t maxOutliers = 0;
+    std::size_t maxSharedHits = 0;
+
+    /// if true, the shared hits are considered as bad hits for this detector
+    bool sharedHitsFlag = false;
+
+    std::size_t detectorId = 0;
+
+    /// a list of values from  0 to 1, the higher number of hits, higher value
+    /// in the list is multiplied to ambuiguity score applied only if
+    /// useAmbiguityFunction is true
+    std::vector<double> factorHits = {1.0};
+
+    /// a list of values from  0 to 1, the higher number of holes, lower value
+    /// in the list is multiplied to ambuiguity score applied only if
+    /// useAmbiguityFunction is true
+    std::vector<double> factorHoles = {1.0};
+  };
+
+  /// @brief  TrackFeatures struct : contains the features that are counted for each track.
+  ///
+  /// The trackFeatures is used to compute the score of each track
+  struct TrackFeatures {
+    std::size_t nHits = 0;
+    std::size_t nHoles = 0;
+    std::size_t nOutliers = 0;
+    std::size_t nSharedHits = 0;
+  };
+
+  /// @brief Configuration struct : contains the configuration for the ambiguity resolution.
+  struct Config {
+    std::map<std::size_t, std::size_t> volumeMap = {{0, 0}};
+    std::vector<DetectorConfig> detectorConfigs;
+    /// minimum score for any track
+    double minScore = 0;
+    /// minimum score for shared tracks
+    double minScoreSharedTracks = 0;
+    /// maximum number of shared tracks per measurement
+    std::size_t maxSharedTracksPerMeasurement = 10;
+    /// maximum number of shared hit per track
+    std::size_t maxShared = 5;
+
+    double pTMin = 0 * UnitConstants::GeV;
+    double pTMax = 1e5 * UnitConstants::GeV;
+
+    double phiMin = -M_PI * UnitConstants::rad;
+    double phiMax = M_PI * UnitConstants::rad;
+
+    double etaMin = -5;
+    double etaMax = 5;
+
+    // if true, the ambiguity score is computed based on a different function.
+    bool useAmbiguityFunction = false;
+  };
+
+  /// @brief OptionalCuts struct : contains the optional cuts to be applied.
+  ///
+  /// The optional cuts,weights and score are used to remove tracks that are not
+  /// good enough, based on some criteria. Users are free to add their own cuts
+  /// with the help of this struct.
+  template <typename track_container_t, typename traj_t,
+            template <typename> class holder_t, bool ReadOnly>
+  struct OptionalCuts {
+    using OptionalFilter =
+        std::function<bool(const Acts::TrackProxy<track_container_t, traj_t,
+                                                  holder_t, ReadOnly>&)>;
```

Comment:
I didn't quiet understand what you meant.   you mean we could directly use `track_container_t::TrackProxy` instead of Acts::TrackProxy ?

Reply:
Instead of templating `template <typename track_container_t, typename traj_t, template <typename> class holder_t, bool ReadOnly>` everywhere you could also just use `template <typename track_container_t>` and the user can point it to `Acts::TrackContainer`. I think that does the same but replaces the various template args with a single one with is the one you use at the end.

It's not so important and can also be followed up later

Reply:
I think I will maybe do it in a different PR as the templating part is bound to change when we change scoring and optional cuts.